examples/visit.rs

```diff
@@ -0,0 +1,250 @@
+//! Example for how to use `VisitMut` to iterate over a table.
+
+use std::collections::BTreeSet;
+use toml_edit::visit::*;
+use toml_edit::visit_mut::*;
+use toml_edit::{Array, Document, Item, Table, Value};
+
+/// This models the visit state for dependency keys in a `Cargo.toml`.
+///
+/// Dependencies can be specified as:
+///
+/// ```toml
+/// [dependencies]
+/// dep1 = "0.2"
+///
+/// [build-dependencies]
+/// dep2 = "0.3"
+///
+/// [dev-dependencies]
+/// dep3 = "0.4"
+///
+/// [target.'cfg(windows)'.dependencies]
+/// dep4 = "0.5"
+///
+/// # and target build- and dev-dependencies
+/// ```
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+enum VisitState {
+    /// Represents the root of the table.
+    Root,
+    /// Represents "dependencies", "build-dependencies" or "dev-dependencies", or the target
+    /// forms of these.
+    Dependencies,
+    /// Represents "target".
+    Target,
+    /// "target.[TARGET]".
+    TargetWithSpec,
+    /// Represents some other state.
+    Other,
+}
+
+impl VisitState {
+    /// Figures out the next visit state, given the current state and the given key.
+    fn descend(self, key: &str) -> Self {
+        match (self, key) {
+            (
+                VisitState::Root | VisitState::TargetWithSpec,
+                "dependencies" | "build-dependencies" | "dev-dependencies",
+            ) => VisitState::Dependencies,
+            (VisitState::Root, "target") => VisitState::Target,
+            (VisitState::Root | VisitState::TargetWithSpec, _) => VisitState::Other,
+            (VisitState::Target, _) => VisitState::TargetWithSpec,
+            (VisitState::Dependencies, _) => VisitState::Dependencies,
+            (VisitState::Other, _) => VisitState::Other,
+        }
+    }
+}
+
+/// Collect the names of every dependency key.
+#[derive(Debug)]
+struct DependencyNameVisitor<'doc> {
+    state: VisitState,
+    names: BTreeSet<&'doc str>,
+}
+
+impl<'doc> Visit<'doc> for DependencyNameVisitor<'doc> {
+    fn visit_table_kv(&mut self, key: &'doc str, node: &'doc Item) {
```

Comment:
Should this also have an implementation for inline tables or assert that dependency tables aren't inline tables?

Might be easy for people looking over this to overlook the inline vs standard stable split

Reply:
btw In other parts of the API we unify the two with TableLike and/or cheat and allow inline tables to be accessed like standard tables (which can allow people to insert an `Item into an inline table which will then get ignored).

Reply:
Hmmmm, good point! Let me see if I can fit `TableLike` into this model as well -- I have some ideas I can play around with.